### PR TITLE
Feature/platfowner/cross shard

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Add implementations for FeatureFlags.enableStateVersionOpt = true
- Implement version APIs of StateNode
- Add tests for state-node.js
- Add more tests for state-util.js

#143

### CODE REVIEW

**Comment by minsulee2** [comment]
Do we not need /get_version?ref=/ stuff?

**Comment by platfowner** [comment]
> Do we not need /get_version?ref=/ stuff? It would be great to see /get_version on REST API as well.

Do you mean finalized version? What would be the use case?

**Comment by platfowner** [comment]
> > Do we not need /get_version?ref=/ stuff? It would be great to see /get_version on REST API as well.
> 
> Do you mean finalized version? What would be the use case?

As versions such node states that change quickly, I tried to log them instead of providing client APIs for them. PTAL

**Comment by minsulee2** [comment]
> > > Do we not need /get_version?ref=/ stuff? It would be great to see /get_version on REST API as well.
> > 
> > 
> > Do you mean finalized version? What would be the use case?
> 
> As versions such node states that change quickly, I tried to log them instead of providing client APIs for them. PTAL

Oh, I usually use REST API for debugging purpose so that, thought it might be necessary for the future! Thanks for the quick updates and seems fare enough :)

**Comment by platfowner** [comment]
> > > > Do we not need /get_version?ref=/ stuff? It would be great to see /get_version on REST API as well.
> > > 
> > > 
> > > Do you mean finalized version? What would be the use case?
> > 
> > 
> > As versions such node states that change quickly, I tried to log them instead of providing client APIs for them. PTAL
> 
> Oh, I usually use REST API for debugging purpose so that, thought it might be necessary for the future! Thanks for the quick updates and seems fare enough :)

Just added /state_versions API. PTAL

File: db/state-node.js [hunk 1]
```diff
@@ -1,24 +1,36 @@
...
     // Used for leaf nodes only.
     this.value = null;
-    this.proof = null;
+    this.proofHash = null;
+    this.version = version ? version : null;
   }
 
-  static create(isLeaf, childMap, value, proof) {
+  static _create(isLeaf, childMap, value, proofHash, version) {
     const node = new StateNode();
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/state-node.js]: Done. Thanks.

File: db/state-node.js [hunk 2]
```diff
@@ -78,36 +87,28 @@ class StateNode {
...
-  hasVersion() {
-    // TODO(lia): Implement this.
+  setProofHash(proofHash) {
+    this.proofHash = proofHash;
   }
 
-  deleteVersion() {
-    // TODO(lia): Implement this.
+  setVersion(version) {
+    return this.version = version;
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/state-node.js]: Done.

File: db/state-node.js [hunk 3]
```diff
@@ -78,36 +87,28 @@ class StateNode {
...
-  getVersions() {
-    // TODO(lia): Implement this.
+  getVersion() {
+    return this.version;
   }
 
-  resetVersions() {
-    // TODO(lia): Implement this.
+  resetVersion() {
+    return this.version = null;
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/state-node.js]: Removed code.

**Review by platfowner — COMMENTED** [review]
All comments addressed. PTAL

File: db/state-node.js [hunk 4]
```diff
@@ -31,12 +42,10 @@ class StateNode {
 
   resetValue() {
```
Comment (liayoo) [inline db/state-node.js:43]: Looks like resetValue() is done only for leaf nodes. Maybe we could check if (this.isLeaf) here and remove from other places?

**Review by liayoo — APPROVED** [review]
LGTM with a comment, thanks! 👍

**Review by minsulee2 — APPROVED** [review]
LGTM

File: db/state-node.js (hunk 4 above)
Comment (platfowner) [reply db/state-node.js:43]: As resetValue() is used in reset() as well (so its purpose is to reset the value anyway), just removed the if condition checks. Thanks.

**Comment by platfowner** [comment]
Thanks for the review, now merging..

**Pull request merged**